zebra-state/src/service/chain_tip.rs

```diff
@@ -410,7 +467,13 @@ impl ChainTipChange {
             // Wait until there is actually Some block,
             // so we don't have `Option`s inside `TipAction`s.
             if let Some(block) = self.best_tip_block() {
-                return Ok(block);
+                // Wait until we have a new block
+                //
+                // last_tip_change() updates last_change_hash, but it doesn't call receiver.changed().
+                // So code that uses both sync and async methods can have spurious pending changes.
+                if Some(block.hash) != self.last_change_hash {
+                    return Ok(block);
+                }
             }
```

Comment:
Here's the actual fix, the rest is tests and instrumentation.

Reply:
As long we're actually updating `last_change_hash`, then `borrow_and_update` would be useful here.

(If we were just peeking, we wouldn't want to mark the value as seen. But I don't think we have any code that just peeks yet.)

Reply:
Added a TODO comment in 5ead11022

Reply:
Should this link to an issue? Either a new issue or maybe the same issue we're using for the Tokio upgrade? For the latter, we could split up the sub-tasks afterwards when we start working on it.